### PR TITLE
Start with system chat messages

### LINKED ISSUE #614: [GSoC idea] Chat system messages
It should be possible for the app to show system messages in the chat.

Samples:
* Person A started a call
* Person B joined the call
* Person B left the call
* Person C was added to the room
* Person A made this room public

Those messages should be stored the same way the chat messages are stored but should display slightly different (less noisy).

### PR DESCRIPTION
**Pull request by nickvergessen**
Fix #614 

### Messages
- [x] Join/leave a call
- [x] Last one leaves: call summary similar to activity
- [x] Add/remove a user
- [x] (Dis-)Allow guests
- [x] Password changes/set/unset
- [x] Promotion/demotion to moderator
- [x] Room created
- [x] Room renamed

### Special design:
* message is 0.5 opacity like date/name on others
* mentions are not bold and not coloured
* no avatar and user name on those rows

> bildschirmfoto von 2018-07-20 17-27-01

### CODE REVIEW

**Comment by nickvergessen** [comment]
Ready for review @Ivansss 

@danxuliu can you help with the layout when the chat is in the sidebar? The timestamp seems to overlap the message there, due to the missing author row parts.

@jancborchardt if you want to comment on the look, do it soon, please, not 2 days before the release ;)

**Comment by danxuliu** [comment]
@nickvergessen I have fixed the date on system messages. The fix itself is the first commit; the second commit provides a nicer layout (at least, in my opinion), but I added it in a separate commit easily droppable just in case I am the only one that likes it ;-)

Original:
system-messages-date-original

Date on its own row (well, in this screenshot the text of the date when there is an author is not right-aligned, but it is fixed in the pushed commit):
system-messages-date-own-row

Date and system messages merged:
system-messages-date-merged

File: docs/api-v1.md [hunk 1]
```diff
@@ -549,6 +549,7 @@ Base endpoint is: `/ocs/v2.php/apps/spreed/api/v1`
 
 * `joined_call` - {actor} joined the call
 * `left_call` - {actor} left the call
+* `call_ended` - Call with {user1}, {user2}, {user3}, {user4} and {user5} (Duration 30:23)
```
Comment (danxuliu) [inline docs/api-v1.md]: Shouldn't it be `Duration {duration}` instead?

Comment (nickvergessen) [reply docs/api-v1.md]: No, similar to the room name on rename its not a placeholder (not a known rich object string type)

**Comment by fancycode** [comment]
The system chat messages should trigger an event to the standalone signaling server, so the clients get notified that they should load the new messages (similar to the regular chat messages).

With #1074 the ChatManager now has a dispatcher that could be used for this.

**Comment by nickvergessen** [comment]
@fancycode after adding the event+listener to update the chat, the signaling backend tests fail because the message is different:
```
1) OCA\Spreed\Tests\php\Signaling\BackendNotifierTest::testRoomInvite
Failed asserting that Array &0 (
    'type' => 'message'
    'message' => Array &1 (
        'data' => Array &2 (
            'type' => 'chat'
            'chat' => Array &3 (
                'refresh' => true
            )
        )
    )
) is identical to Array &0 (
    'type' => 'invite'
    'invite' => Array &1 (
        'userids' => Array &2 (
            0 => 'testUser'
        )
        'alluserids' => Array &3 (
            0 => 'testUser'
        )
        'properties' => Array &4 (
            'name' => ''
            'type' => 3
        )
    )
).
```

Can you help with that?

**Comment by fancycode** [comment]
> Can you help with that?

The tests only remember the last message that was received: https://github.com/nextcloud/spreed/blob/bd788511aa599cf80d0d1ce4640ded09afc46a9e/tests/php/Signaling/BackendNotifierTest.php#L51

With the system messages there will be more now: first the `invite` that would be checked by the test, then the `refresh` triggered by the system chat. You could either filter the `chat` messages in the test, or keep a list of messages and check if the expected message is in the list.

File: tests/php/Signaling/BackendNotifierTest.php [hunk 2]
```diff
@@ -37,36 +37,36 @@
...
-        return $this->lastRequest;
-    }
+	public function getRequests() {
+		return $this->requests;
+	}
 
-    public function clearLastRequest() {
-        $this->lastRequest = null;
-    }
+	public function clearLastRequest() {
```
Comment (fancycode) [inline tests/php/Signaling/BackendNotifierTest.php]: Please rename to `clearRequests` to match the other function names.

**Pull request merged**